x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java

```diff
@@ -269,17 +274,74 @@ public boolean equals(Object obj) {
             TopMetric other = (TopMetric) obj;
             return sortFormat.equals(other.sortFormat)
                     && sortValue.equals(other.sortValue)
-                    && Arrays.equals(metricValues, other.metricValues);
+                    && metricValues.equals(other.metricValues);
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(sortFormat, sortValue, Arrays.hashCode(metricValues));
+            return Objects.hash(sortFormat, sortValue, metricValues);
         }
 
         @Override
         public String toString() {
-            return "TopMetric[" + sortFormat + "," + sortValue + "," + Arrays.toString(metricValues) + "]"; 
+            return "TopMetric[" + sortFormat + "," + sortValue + "," + metricValues + "]"; 
+        }
+    }
+
+    static class MetricValue implements Writeable, ToXContent {
+        private final DocValueFormat format; 
+        private final SortValue value;
```

Comment:
I'm aware it is odd to have a "SortValue" be part of a MetricValue. It is a very convenient way to send a type-aware thing across the wire though.

Reply:
I think this would be better as an inline comment than a PR comment.

Reply:
:+1: